rebalance/rebalance.py

```diff
@@ -212,16 +215,260 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi=None,
+            continue
+        # We might have multiple channel entries ! Eg if one was just closed
+        # and reopened.
+        for chan in peer["channels"]:
+            if "short_channel_id" not in chan:
+                continue
+            if chan["short_channel_id"] == scid:
+                return chan
+
+
+def maybe_rebalance_pairs(plugin: Plugin, ch1, ch2, failed_pairs: list):
+    scid1 = ch1["short_channel_id"]
+    scid2 = ch2["short_channel_id"]
+    result = {"success": False, "fee_spent": 0}
+    if scid1 + ":" + scid2 in failed_pairs:
+        return result
+    min_amount = int(plugin.min_amount)
+    i = 0
+    while not plugin.rebalance_stop:
+        amount1 = min(must_send(ch1, plugin.enough_liquidity),
+                      could_receive(ch2, plugin.enough_liquidity))
+        amount2 = min(should_send(ch1, plugin.enough_liquidity, plugin.ideal_ratio),
+                      should_receive(ch2, plugin.enough_liquidity, plugin.ideal_ratio))
+        amount3 = min(could_send(ch1, plugin.enough_liquidity),
+                      must_receive(ch2, plugin.enough_liquidity))
+        amount = max(amount1, amount2, amount3)
+        if amount < min_amount:
+            return result
+        amount = int(min(amount, get_max_amount(i, min_amount, plugin.enough_liquidity)))
+        maxfee = get_max_fee(plugin, amount)
+        plugin.log(f"Try to rebalance: {scid1} -> {scid2}; amount={amount:,} msat; maxfee={maxfee:,} msat")
+        try:
+            res = rebalance(plugin, outgoing_scid=scid1, incoming_scid=scid2, msatoshi=amount, maxfeepercent=0, retry_for=1200, exemptfee=maxfee)
+        except Exception:
+            failed_pairs.append(scid1 + ":" + scid2)
+            # rebalance failed, let's try with a smaller amount
+            while (get_max_amount(i, min_amount, plugin.enough_liquidity) >= amount and
+                   get_max_amount(i, min_amount, plugin.enough_liquidity) != get_max_amount(i + i, min_amount, plugin.enough_liquidity)):
+                i += 1
+            if amount > get_max_amount(i, min_amount, plugin.enough_liquidity):
+                continue
+            return result
+        plugin.log(f"Rebalance succeeded: {res}")
+        result["success"] = True
+        result["fee_spent"] += res["fee"]
+        # refresh channels
+        time.sleep(10)
+        ch1 = get_chan(plugin, scid1)
+        assert ch1 is not None
+        ch2 = get_chan(plugin, scid2)
+        assert ch2 is not None
+
+
+def maybe_rebalance_once(plugin: Plugin, failed_pairs: list):
+    channels = get_our_channels(plugin)
+    for ch1 in channels:
+        for ch2 in channels:
+            result = maybe_rebalance_pairs(plugin, ch1, ch2, failed_pairs)
+            if result["success"] or plugin.rebalance_stop:
+                return result
+    return {"success": False, "fee_spent": 0}
+
+
+def rebalanceall_thread(plugin: Plugin):
+    try:
+        channels = get_our_channels(plugin)
```

Comment:
First line here should actually do the `if not plugin.mutex.acquire(blocking=False): ...` acquire lock check, so we have the locks ownership on the correct thread and not the starter function.